### PR TITLE
Feature/mailbox creation improvement

### LINKED ISSUE #278: Formulaire de création de boîte mail : indiquer les champs obligatoires en début de formulaire
Important pour l'UX et pour l'accessibilité.

Si tous les champs restent obligatoires : on ajoute la phrase ci-dessous en haut du formulaire.
FR : **Tous les champs sont obligatoires.**
EN : **All fields are mandatory.**

### PR DESCRIPTION
**Pull request by daproclaima**
## Purpose

Improve UI and a11y for mailbox creation form.
- Quotes that all fields are mandatory;
- Quotes email address prefix syntax to respect;
- API errors related to invalid data sent from form fields (wrong local part, already used local part, wrong secondary email address) are indicated on the field, not above the form;

Ticket related to issue #281.

## Proposal

- [x] shows "All fields are mandatory." above the form;
- [x] all fields have an aria-required=true attribute;
- [x] mail email address field has the description 'do not include "@" in it';
- [x] API error causes such as "Enter a valid value.", "Enter a valid email address.", "Mailbox with this Local_part and Domain already exists." are not shown anymore above the form but are used to highlight the corresponding invalid field. Other errors are still displayed above the form;

### Before
before

### After
after1
after2

close #278, #279

### CODE REVIEW

File: src/frontend/apps/desk/src/features/mail-domains/components/forms/CreateMailboxForm.tsx [hunk 1]
```diff
@@ -157,6 +234,7 @@ const Form = ({
             render={({ fieldState }) => (
               <Input
                 aria-invalid={!!fieldState.error}
+                aria-required="true"
```
Comment (AntoLC) [inline src/frontend/apps/desk/src/features/mail-domains/components/forms/CreateMailboxForm.tsx:237]: You can just used the input properties, you don't need all this check I think.


```suggestion
                aria-required="true"
                required
```

File: src/frontend/apps/desk/src/features/mail-domains/components/forms/CreateMailboxForm.tsx [hunk 2]
```diff
@@ -177,6 +255,7 @@ const Form = ({
...
```
Comment (AntoLC) [inline src/frontend/apps/desk/src/features/mail-domains/components/forms/CreateMailboxForm.tsx:258]:
```suggestion
                aria-required="true"
                required
```

File: src/frontend/apps/desk/src/features/mail-domains/components/forms/CreateMailboxForm.tsx [hunk 3]
```diff
@@ -198,14 +277,23 @@ const Form = ({
               render={({ fieldState }) => (
                 <Input
                   aria-invalid={!!fieldState.error}
+                  aria-required="true"
```
Comment (AntoLC) [inline src/frontend/apps/desk/src/features/mail-domains/components/forms/CreateMailboxForm.tsx:280]:
```suggestion
                  aria-required="true"
                  required
                  pattern="^((?!@).)*$"
```

File: src/frontend/apps/desk/src/features/mail-domains/components/forms/CreateMailboxForm.tsx [hunk 4]
```diff
@@ -236,10 +324,17 @@ const Form = ({
             render={({ fieldState }) => (
               <Input
                 aria-invalid={!!fieldState.error}
+                aria-required="true"
```
Comment (AntoLC) [inline src/frontend/apps/desk/src/features/mail-domains/components/forms/CreateMailboxForm.tsx:327]:
```suggestion
                aria-required="true"
                required
                pattern="[^@\s]+@[^@\s]+\.[^@\s]+"
```

File: src/frontend/apps/e2e/__tests__/app-desk/mail-domain-create-mailbox.spec.ts [hunk 5]
```diff
@@ -269,4 +289,254 @@ test.describe('Mail domain create mailbox', () => {
...
+
+    await clickOnMailDomainsNavButton(page);
+
+    await page.getByRole('listbox').first().getByText('domain.fr').click();
+
+    await page.getByRole('button', { name: 'Create a mailbox' }).click();
+    await page.getByRole('button', { name: 'Cancel' }).click();
+
+    await expect(page.getByTitle('Mailbox creation form')).toBeHidden();
+
```
Comment (AntoLC) [inline src/frontend/apps/e2e/__tests__/app-desk/mail-domain-create-mailbox.spec.ts:381]: It doesn't seem related to this test.

```suggestion
```

File: src/frontend/apps/e2e/__tests__/app-desk/mail-domain-create-mailbox.spec.ts [hunk 6]
```diff
@@ -269,4 +289,254 @@ test.describe('Mail domain create mailbox', () => {
...
+    const inputMainEmailAddress = page.getByLabel('Main email address');
+    const inputSecondaryEmailAddress = page.getByLabel(
+      'Secondary email address',
+    );
+    const submitButton = page.getByRole('button', { name: 'Submit' });
+
+    await inputFirstName.fill('John');
+    await inputLastName.fill('Doe');
+    await inputMainEmailAddress.fill('wrong@');
+    await expect(page.getByText('do not include "@" in it')).toBeVisible();
```
Comment (AntoLC) [inline src/frontend/apps/e2e/__tests__/app-desk/mail-domain-create-mailbox.spec.ts:400]: Try to check only what this test is about, I have the feeling that it is the same assertion has the previous test.

File: src/frontend/apps/e2e/__tests__/app-desk/mail-domain-create-mailbox.spec.ts [hunk 7]
```diff
@@ -269,4 +289,254 @@ test.describe('Mail domain create mailbox', () => {
...
+                },
+              });
+            }
+
+            countCreateMailboxRequestSent++;
+          }
+        },
+        { times: 2 },
+      );
+    };
```
Comment (AntoLC) [inline src/frontend/apps/e2e/__tests__/app-desk/mail-domain-create-mailbox.spec.ts:369]: It makes the test very heavy, it is not possible to make it DRY ?

File: src/frontend/apps/e2e/__tests__/app-desk/mail-domain-create-mailbox.spec.ts [hunk 8]
```diff
@@ -269,4 +289,254 @@ test.describe('Mail domain create mailbox', () => {
...
+
+    await clickOnMailDomainsNavButton(page);
+
+    await page.getByRole('listbox').first().getByText('domain.fr').click();
+
+    await page.getByRole('button', { name: 'Create a mailbox' }).click();
+    await page.getByRole('button', { name: 'Cancel' }).click();
+
+    await expect(page.getByTitle('Mailbox creation form')).toBeHidden();
+
```
Comment (AntoLC) [inline src/frontend/apps/e2e/__tests__/app-desk/mail-domain-create-mailbox.spec.ts:511]: Same remark as before, don't assert things that are already asserted.

**Review by AntoLC — COMMENTED** [review]
I have the feeling this component is getting very complicated for something that could be handle easily. 
You can do some checkup without too much code directly with html (props `required` / `pattern`), by using them you will not need to have complicated state or text props:
https://github.com/numerique-gouv/people/blob/16e4ad811b4a37823c41eef8b0027f90f5572737/src/frontend/apps/desk/src/features/mail-domains/components/forms/CreateMailboxForm.tsx#L282-L297

See how it react with this props:

[scrnli_7_8_2024_10-05-10 AM.webm](https://github.com/numerique-gouv/people/assets/25994652/519d8cc4-2813-49c7-a3de-bcb5e38d2819)

File: src/frontend/apps/desk/src/features/mail-domains/components/forms/CreateMailboxForm.tsx [hunk 9]
```diff
@@ -57,26 +85,38 @@ export const CreateMailboxForm = ({
...
   const onSubmitCallback = (event: React.FormEvent) => {
     event.preventDefault();
     void methods.handleSubmit((data) =>
       createMailbox({ ...data, mailDomainSlug: mailDomain.slug }),
     )();
   };
 
+  const causes = error?.cause?.filter((cause) => {
+    const isFound =
+      cause === 'Mailbox with this Local_part and Domain already exists.';
```
[the remark this answers is not in the document]
Comment (daproclaima) [reply src/frontend/apps/desk/src/features/mail-domains/components/forms/CreateMailboxForm.tsx:108]: Exactly. We will discuss this on ticket #281.

**Review by AntoLC — APPROVED** [review]
👌
